python/cudf/cudf/core/reshape.py

```diff
@@ -353,35 +387,67 @@ def concat(objs, axis=0, join="outer", ignore_index=False, sort=None):
             objs = _align_objs(objs, how=join, sort=sort)
             df.index = objs[0].index
 
-        for o in objs:
-            for name, col in o._data.items():
-                if name in df._data:
-                    raise NotImplementedError(
-                        f"A Column with duplicate name found: {name}, cuDF "
-                        f"doesn't support having multiple columns with "
-                        f"same names yet."
-                    )
-                if empty_inner:
-                    # if join is inner and it contains an empty df
-                    # we return an empty df, hence creating an empty
-                    # column with dtype metadata retained.
-                    df[name] = cudf.core.column.column_empty_like(
-                        col, newsize=0
-                    )
-                else:
-                    df[name] = col
+        # if the dictionary consists of only dictionaries
+        # it must be handled differently
+        only_series = len(typs) == 1 and cudf.Series in typs
 
-        result_columns = (
-            objs[0]
-            ._data.to_pandas_index()
-            .append([obj._data.to_pandas_index() for obj in objs[1:]])
-        )
+        if keys is None:
+            for o in objs:
+                for name, col in o._data.items():
+                    if name in df._data:
+                        raise NotImplementedError(
+                            f"A Column with duplicate name found: {name}, cuDF "
+                            f"doesn't support having multiple columns with "
+                            f"same names yet."
+                        )
+                    if empty_inner:
+                        # if join is inner and it contains an empty df
+                        # we return an empty df, hence creating an empty
+                        # column with dtype metadata retained.
+                        df[name] = cudf.core.column.column_empty_like(
+                            col, newsize=0
+                        )
+                    else:
+                        df[name] = col
+
+            result_columns = (
+                objs[0]
+                ._data.to_pandas_index()
+                .append([obj._data.to_pandas_index() for obj in objs[1:]])
+                .unique()
+            )
 
-        if ignore_index:
-            # with ignore_index the column names change to numbers
-            df.columns = pd.RangeIndex(len(result_columns.unique()))
+        # need to create a MultiIndex column
         else:
+            for k, o in zip(keys, objs):
+                for name, col in o._data.items():
+                    # if only series, then only keep keys as column labels
+                    # if the existing column is multiindex, prepend it
+                    # to handle cases where dfs and srs are concatenated,
+                    # explicitly cast int column labels into str
```

Comment:
*question*: This cast is required because each level of a cudf multiindex must have a homogeneous type?

If so, one issue with this is that this diverges from pandas (which will put heterogeneous types into a single level).

My feeling is that we should _probably_ gate this cast behind cudf not being in "pandas-compatibility" mode, and instead raise if we are. cc @galipremsagar for a cudf.pandas opinion.

Reply:
we're using `pyarrow` to convert the data into an array. with heterogeneous types, `pyarrow` tries to convert the types to be the same type -- hence why amanlai was converting the data to `str` before it'd get to `pyarrow`.

will wait for another maintainer to chime in here on best approach, i'm too new to have context

Reply:
cudf columns, like pyarrow, only allow for homogeneous dtypes. In contrast, pandas has type-specialised columns, but a python `object` catchall as a fallback.

In pandas-compatibility mode, we want to replicate (as closely as possible) pandas behaviour and explicitly raise if we can't do so. So that would be:

```python
if cudf.get_option("mode.pandas_compatible"):
   raise NotImplementedError(some useful error message)
```

In non-compat mode we have a bit more freedom, so we could cast to string. _But_ my preference is, where possible, to maintain fidelity of the user input and raise where that is not possible: that would mean not casting to string and also raising in non-compat mode.

Reply:
this approach makes sense to me, and perhaps we can look into type-specialized columns as a later PR

Reply:
i've raised the error regardless of `"mode.pandas_compatible"`.

i looked into creating an `option` to allow the user to override this behavior and cast to `str` anyway, but based on what i'm seeing in `options.py` we try to keep the `option`s few in number, so i won't do that for now. let me know if you'd like me to create an option to override.

Reply:
i've also removed the cast to `str` so the column labels match `pandas` output